### PR TITLE
gh_actions: configure to deploy using Github Pages

### PR DESCRIPTION
**Pull request by leandrolanzieri**
## Contribution description
This changes the Github workflow to build and deploy the website to the `gh-pages` branch. You can see the actions working [here](https://github.com/leandrolanzieri/riot-os.org-1/actions), and the result here: https://leandrolanzieri.github.io/riot-os.org-1/ (note that as my repo has a different name, an extra commit to change the base URL had to be added).

Some fixes were also added to use `baseurl` throughout the code.

After this, we still need to change the Github Pages configuration to get the files from the `gh-pages` branch.

### CODE REVIEW

**Comment by leandrolanzieri** [comment]
Maybe @aabadie wants to take a look as well?

**Review by miri64 — COMMENTED** [review]
LGTM, but I leave the final word to @aabadie since he is the real GH actions pro ;-)

File: README.md [hunk 1]
```diff
@@ -8,6 +8,10 @@ production environment set the environmental variable `JEKYLL_ENV=production`.
 To watch for file changes and trigger a rebuild, you can run
 `jekyll build --watch`.
 
+For local environments you need to add the `_config_development.yml` file to the
+build to override the `baseurl`, by adding `--config _config.yml,_config_development.yml` to
```
Comment (aabadie) [inline README.md:12]: What `baseurl` is supposed to contain in a development environment ? `/` ? Might be worth documenting this.

File: .github/workflows/deployment.yml [hunk 2]
```diff
@@ -47,14 +47,15 @@ jobs:
...
           printf "contributors: %d\n" ${{steps.fetch-contributors.outputs.contributors-numof}} >> ${{ github.workspace }}/_data/riot_stats.yml
-      - name: Build
-        uses: jerryjvl/jekyll-build-action@v1
+      # Use GitHub Actions' cache to shorten build times and decrease load on servers
+      - uses: actions/cache@v2
+        with:
+          path: vendor/bundle
+          key: ${{ runner.os }}-gems-${{ hashFiles('**/Gemfile') }}
+          restore-keys: |
+            ${{ runner.os }}-gems-
```
Comment (aabadie) [inline .github/workflows/deployment.yml:56]:
```suggestion
          restore-keys: ${{ runner.os }}-gems-
```
?

File: README.md (hunk 1 above)
Comment (leandrolanzieri) [reply README.md:12]: I added the `_config_development.yml` file, it's an empty string

File: .github/workflows/deployment.yml [hunk 3]
```diff
@@ -47,6 +47,13 @@ jobs:
         run: |
           echo '${{steps.fetch-contributors.outputs.result}}' > ${{ github.workspace }}/_data/contributors.json
           printf "contributors: %d\n" ${{steps.fetch-contributors.outputs.contributors-numof}} >> ${{ github.workspace }}/_data/riot_stats.yml
+      # Use GitHub Actions' cache to shorten build times and decrease load on servers
+      - uses: actions/cache@v2
+        with:
+          path: vendor/bundle
+          key: ${{ runner.os }}-gems-${{ hashFiles('**/Gemfile') }}
+          restore-keys: |
+            ${{ runner.os }}-gems-
```
Comment (aabadie) [inline .github/workflows/deployment.yml]: All this change is part of an unrelated commit (the one about the base url)

Comment (leandrolanzieri) [reply .github/workflows/deployment.yml]: Right, should I split it into its own or merge it with 0fb5aa96e7773c8bcc20d86b19fb76cc7454fb7e ?

Comment (aabadie) [reply .github/workflows/deployment.yml]: I don't know. Is caching really needed for such a small site ? What is the performance gain ?

Comment (leandrolanzieri) [reply .github/workflows/deployment.yml]: When I first tried it it improved compared to not having it, I think it dropped from ~4 minutes to less than 2 minutes deploy time. It is also used in the example of the jekyll action (https://github.com/helaili/jekyll-action).

**Comment by aabadie** [comment]
Thinking about the workflow and there's something I don't like: all fresh statistics (list of boards, contributors, etc) are only computed by CI.
Do you think the Jekyll hooks could be used instead to generate all these information during the build and not by preliminary steps ?
That would allow developers to get the same content and rendering locally and to limit the dependency to github actions.

**Comment by leandrolanzieri** [comment]
> Thinking about the workflow and there's something I don't like: all fresh statistics (list of boards, contributors, etc) are only computed by CI.
> Do you think the Jekyll hooks could be used instead to generate all these information during the build and not by preliminary steps ?
> That would allow developers to get the same content and rendering locally and to limit the dependency to github actions.

I agree it would be nice to have the fresh stats locally as well. But does it mean that for building this the RIOT repo would also be cloned? How would it look like?

**Comment by waehlisch** [comment]
> > Thinking about the workflow and there's something I don't like: all fresh statistics (list of boards, contributors, etc) are only computed by CI.
> > Do you think the Jekyll hooks could be used instead to generate all these information during the build and not by preliminary steps ?
> > That would allow developers to get the same content and rendering locally and to limit the dependency to github actions.
> 
> I agree it would be nice to have the fresh stats locally as well. But does it mean that for building this the RIOT repo would also be cloned? How would it look like?

i don't see a  necessary requirement. i may imagine that website editors become annoyed having the full code base.

**Comment by waehlisch** [comment]
if no further actions are required, i think we can proceed and merge.

**Comment by aabadie** [comment]
> But does it mean that for building this the RIOT repo would also be cloned? How would it look like?

It could use a RIOT submodule that is pointing to master => each time the website is rebuilt, the submodule is updated to fetch the latest master version of RIOT.

**Review by waehlisch — APPROVED** [review]

File: .github/workflows/deployment.yml (hunk 3 above)
Comment (aabadie) [reply .github/workflows/deployment.yml]: This comment still needs to be addressed. I think this change should be in its own commit.

(And this PR needs squashing BTW)

Comment (leandrolanzieri) [reply .github/workflows/deployment.yml]: Will do

**Comment by waehlisch** [comment]
> It could use a RIOT submodule that is pointing to master => each time the website is rebuilt, the submodule is updated to fetch the latest master version of RIOT.

let's discuss this in a separate PR. the current version is the minimal working solution.

File: .github/workflows/deployment.yml (hunk 3 above)
Comment (leandrolanzieri) [reply .github/workflows/deployment.yml]: Split in edbc5d6

Comment (waehlisch) [reply .github/workflows/deployment.yml]: +1 for caching given the performance numbers.

**Comment by leandrolanzieri** [comment]
The cache step configuration has its own commit now, and the suggestion commit has been squashed.

**Comment by leandrolanzieri** [comment]
There it goes ! https://riot-os.github.io/riot-os.org/
